Update quanto blogpost (#2246)
huggingface/blog: quanto-introduction.md

---

> title: "Quanto: a pytorch quantization backend for optimum"

Reviewer: ```suggestion
title: "Quanto: a PyTorch quantization backend for Optimum"
```

thumbnail: /blog/assets/169_quanto_intro/thumbnail.png
authors:
- user: dacorvo
- user: ybelkada
- user: marcsun13
---

> # Quanto: a pytorch quantization backend for optimum

Reviewer: ```suggestion
# Quanto: a PyTorch quantization backend for Optimum
```

Also, although they are based on the same design principles, they are unfortunately often incompatible with one another.

> Today, we are excited to introduce [quanto](https://github.com/huggingface/optimum-quanto), a pytorch quantization backend for [Optimum](https://huggingface.co/docs/optimum/index).

Reviewer: ```suggestion
Today, we are excited to introduce [quanto](https://github.com/huggingface/optimum-quanto), a PyTorch quantization backend for [Optimum](https://huggingface.co/docs/optimum/index).
```

It has been designed with versatility and simplicity in mind:

- all features are available in eager mode (works with non-traceable models),
- quantized models can be placed on any device (including CUDA and MPS),
- automatically inserts quantization and dequantization stubs,
- automatically inserts quantized functional operations,
- automatically inserts quantized modules (see below the list of supported modules),

> - provides a seamless workflow from a float model to a dynamic to a static quantized model,

Reviewer: ```suggestion
- provides a seamless workflow from a float model to a dynamic and static quantized model,
```

Author: actually it is dynamic, then static

> - serialization compatible with pytorch `weight_only` and 🤗 `safetensors`,

Reviewer: ```suggestion
- serialization compatible with PyTorch `weight_only` and 🤗 [Safetensors](https://huggingface.co/docs/safetensors/index),
```

- accelerated matrix multiplications on CUDA devices (int8-int8, fp16-int4, bf16-int8, bf16-int4),
- supports int2, int4, int8 and float8 weights,
- supports int8 and float8 activations.

Recent quantization methods appear to be focused on quantizing Large Language Models (LLMs), whereas [quanto](https://github.com/huggingface/optimum-quanto) intends to provide extremely simple quantization primitives for simple quantization schemes (linear quantization, per-group quantization) that are adaptable across any modality.

## Quantization workflow

Quanto is available as a pip package.

```sh
pip install optimum-quanto
```

A typical quantization workflow consists of the following steps:

**1. Quantize**

The first step converts a standard float model into a dynamically quantized model.

```python
from optimum.quanto import quantize, qint8

quantize(model, weights=qint8, activations=qint8)
```

At this stage, only the inference of the model is modified to dynamically quantize the weights.

**2. Calibrate (optional if activations are not quantized)**

> Quanto supports a calibration mode that allows to record the activation ranges while passing representative samples through the quantized model.

Reviewer: ```suggestion
Quanto supports a calibration mode that allows the recording of the activation ranges while passing representative samples through the quantized model.
```

**4. Freeze integer weights**

When freezing a model, its float weights are replaced by quantized weights.

```python
from optimum.quanto import freeze

freeze(model)
```

**5. Serialize quantized model**

Quantized models weights can be serialized to a `state_dict`, and saved to a file.
Both `pickle` and `safetensors` (recommended) are supported.

```python
from safetensors.torch import save_file

save_file(model.state_dict(), 'model.safetensors')
```

> In order to be able to reload these weights, you also need to store the quantized
> model quantization map.

Reviewer: ```suggestion
In order to reload these weights, you also need to store the quantized
models quantization map.
```

```python
import json

from optimum.quanto import quantization_map

with open('quantization_map.json', w) as f:
  json.dump(quantization_map(model))
```

**5. Reload a quantized model**

A serialized quantized model can be reloaded from a `state_dict` and a `quantization_map` using the `requantize` helper.

> Note that you need first to instantiate an empty model.

Reviewer: ```suggestion
Note that you need to first instantiate an empty model.
```
